Add FSDP arguments and example script to train model with FSDP-v2

ryang-amd opened the pull request:
Hi,
cc @wenchenvincent @gurpreet-dhami 

I updated the README file with fsdp-related argument and example commands.
Meanwhile also included the training script for training Llama2 with FSDP-v2 enabled.

Please review and let me know if there needs more explanations.

ryang-amd committed:
add argument and example for running megatron-lm with fsdpv2

ryang-amd committed:
modified README and data directory in the script

ryang-amd committed:
small amendments for  the key argument

wenchenvincent commented on examples/llama/train_llama2_fsdpv2.sh:
```diff
@@ -0,0 +1,342 @@
+#!/bin/bash
```
Can we modify the existing train_llama2.sh and train_llama3.sh instead of adding a new script?

gurpreet-dhami replied:
@wenchenvincent  : We are using llama2 tokenizer,  whereas new fsdp script is using HF tokenizer.
May be we can add an argument as TOKENIZER_TYPE = TOKENIZER_HF  or TOKENIZER_LLAMA2

ryang-amd replied:
working on it

lcskrishna commented on examples/llama/train_llama2_fsdpv2.sh:
```diff
@@ -0,0 +1,342 @@
+  ds_works=24
+fi
+
+if [[ $MODEL_SIZE -eq 8 ]]; then #llama3.1-8B
```
Looks like this is LLama 3.1 configuration. Can you please confirm and update your script name accordingly.

ryang-amd commented:
Hi @wenchenvincent @gurpreet-dhami @lcskrishna 
I've updated the existing scripts and README:
 - add FSDP-v2 parameter settings into train_llama2.sh and train_llama3.sh
 - set `TOKENIZER_TYPE` in llama2 training script to choose different types of tokenizer
 - update the script's config name for different models
 I also tested locally and the two scripts are working well.

Please review.

wenchenvincent commented on examples/llama/readme.md:
````diff
@@ -71,30 +71,35 @@ You can use either mock data or real data for training.
   ```bash
-  DATA_PATH=${DATA_DIR}/wikipedia_20220301.en/wikipedia_20220301.en.train.jsonl_text_document
-  ``` 
+  export HF_TOKEN="hf_xxxx"
````
Could you give examples for xxxx?

ryang-amd replied:
Added instructions for HF_TOKEN in README

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -6,19 +6,20 @@
 export HSA_NO_SCRATCH_RECLAIM=1
-
+export NVTE_DEBUG=1
+export NVTE_DEBUG_LEVEL=2
```
Do we need these two env vars?

ryang-amd replied:
Those are removed in the latest commit.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -36,73 +37,89 @@ TIME_STAMP=$(date +"%Y-%m-%d_%H-%M-%S")
 if [ "$GEMM_TUNING" -eq 1 ]; then
-   export TE_HIPBLASLT_TUNING_RUN_COUNT=10
-   export TE_HIPBLASLT_TUNING_ALGO_COUNT=50
+    export TE_HIPBLASLT_TUNING_RUN_COUNT=20
```
Nit: remove the extra space

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -36,73 +37,89 @@ TIME_STAMP=$(date +"%Y-%m-%d_%H-%M-%S")
 USE_FLASH_ATTN="${USE_FLASH_ATTN:-1}"
 NO_TRAINING="${NO_TRAINING:-0}" # NO_TRAINING=1: for computing metrics only
-ENABLE_PROFILING="${ENABLE_PROFILING:-0}" #enable pytorch profiling
+ENABLE_PROFILING="${ENABLE_PROFILING:-0}"
```
I suggest you keep the comment.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -36,73 +37,89 @@ TIME_STAMP=$(date +"%Y-%m-%d_%H-%M-%S")
 fi
 
 MODEL_SIZE="${MODEL_SIZE:-70}"
-TP="${TP:-8}"
```
When FSDP is enabled, is TP automatically disabled in Megatron? If not, we might want to add checks in this script to make sure the TP size is 1 when FSDP is enabled.

ryang-amd replied:
Added notes in the README file.

ryang-amd replied:
Just checked about TP=8 setting. There seems to be some issue when TP=8 enabled and will raise error. I'll fix this latest tomorrow morning (Finnish time).

wenchenvincent replied:
I saw that you added some code to enable SP when TP is on, even when FSDP is on. My main question is, can we really use TP with FSDP?

ryang-amd replied:
Yes, they can be used together. But it probably requires some careful set ups. We haven't investigated deeper here. In 1 node's testing, when using FSDP+TP it usually degrades the performance. 
It probably worth also trying with multi-node when setting FSDP with hybrid sharding strategy. As for 1 node, changing to different FSDP sharding strategies doesn't provide much performance boost.

wenchenvincent replied:
I would suggest disabling FSDP+TP if we are not sure about the performance. The motivation of using FSDP is to address the impact of the communication overhead caused by TP. I don't quite expect that they would be used together.

ryang-amd replied:
> I would suggest disabling FSDP+TP if we are not sure about the performance. The motivation of using FSDP is to address the impact of the communication overhead caused by TP. I don't quite expect that they would be used together.

Hi @wenchenvincent ,
I've updated accordingly to make FSDP not working together with TP.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -224,10 +268,6 @@ if [ "$USE_FLASH_ATTN" -eq 1 ]; then
 fi
 
-if [ "$SEQ_PARALLEL" -eq 1 ]; then
-EXTRA_ARGS="$EXTRA_ARGS --sequence-parallel"
```
I think we still want to control sequence-parallel via SEQ_PARALLEL

ryang-amd replied:
SEQ_PARALLEL is kept and determined when FSDP is enabled.

wenchenvincent reviewed (changes requested):
@ryang-amd Thanks for the timely updates. I added some more specific comments that. And I have a general suggestion. When you change the existing train_llama2.sh and train_llama3.sh, please make minimal changes and keep the existing default value for those parameters. For the parameters that you need to override, override it through command line and add the command options in the README.

ryang-amd committed:
revert to previous version and apply changes according to instructions

ryang-amd commented:
Hi @wenchenvincent,
Thanks for reviewing and giving me suggestions! 
I've made changes according to your instructions. 
- Basically, train_llama2.sh and train_llama3.sh should now have the minimal changes compared to the original ones. 
- I've also added more instructions in the README file for setting FSDP and downloading Llama3.1 tokenizer.

Please review when it's convenient for you. Thanks!

wenchenvincent commented on examples/llama/readme.md:
```diff
@@ -9,6 +9,8 @@ This guide provides the steps for setting up the environment and configuring the
    Download the Docker image required for training:  
    `docker pull <image_name>`
+   
+   **Note:** FSDP-v2 requires `PyTorch >= 2.4.0` with FSDP-v2 support.
```
@jithunnair-amd I think you mentioned FSDPv2 needes Pytorch 2.5. I did see that Pytorch release notes say it has fdsp v2 in 2.4. Could you clarify?

wenchenvincent replied:
FSDP v2 was first introduced in Pytorch 2.4 but many important features and performance improvement was added in Pytorch 2.5. So let's recommend using Pytorch 2.5 and later for this reason.

ryang-amd replied:
updated in the Readme

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -156,17 +172,35 @@ GPT_ARGS="
     --bf16 \
     --no-masked-softmax-fusion \
     --disable-bias-linear \
+    --no-rope-fusion \
```
Is this due to the issue with the RoPE fusion?

wenchenvincent replied:
I did remember without rope fusion, more memory will be used and we could only use smaller batches. The previous tests were run with rope fusion. I suggest that you put in an argument to specify using rope fusion or not. And by default use rope fusion.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -156,17 +172,35 @@ GPT_ARGS="
 "
 
+
+if [ "$OPTIMIZER" == "adam" ]; then
```
Are we using the distributed optimizer? One reason why we hardcoded using Adam was that the distributed optimizer only supports Adam. So if we use distributed optimizer, we should only use Adam.

ryang-amd replied:
No, when FSDP is enabled, we don't use distributed optimizer (and also disable overlap-param-gather), As it may cause parameter update conflicts (due to two different synchronization and update mechanisms).

But when FSDP is disabled, use-distributed-optimizer and overlap-param-gather are automatically set on. I will add check here to determine which optimizer to use.

In previous my experiments with FSDP on Llama3, SGD usually consumes less memory and allows larger MBS,  compared to Adam.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -177,9 +211,9 @@ DATA_ARGS="
     --eval-iters 10 \
     --num-workers $ds_works \
-    --mock-data
+    --data-path $DATA_PATH \
```
Could you also add an option to let the user either use mock data or the real data?

ryang-amd commented on examples/llama/train_llama2.sh:
```diff
@@ -76,19 +76,39 @@ CONTI_PARAMS="${CONTI_PARAMS:-0}"
+    echo "Using HuggingFaceTokenizer."
 fi
 
 MAX_POSITION_EMBEDDINGS=4096
```
@wenchenvincent Do you know why here the default value is set to be 4096? It will not work with seq_length=8192 which is greater than 4096. Can I change it to a larger number in the script?

wenchenvincent replied:
@lizamd Do you know why?

wenchenvincent replied:
@ryang-amd What's the difference between `MAX_POSITION_EMBEDDINGS` and `SEQ_LENGTH`? I see that `MAX_POSITION_EMBEDDINGS` is used for `max_sequence_length` in the model.

ryang-amd replied:
`MAX_POSITION_EMBEDDINGS ` determines the size of the position embedding in the model, which affects model's ability to process longer sequences. SEQ_LENGTH is the actual sequence length used during training, it must be <= `MAX_POSITION_EMBEDDINGS`, otherwise extra tokens from the sequence will not have corresponding position embeddings. `max_sequence_length` is usually set to `MAX_POSITION_EMBEDDINGS` to ensure that each token in the sequence has a unique position, as the position embeddings define how many unique token positions the model can distinguish.

wenchenvincent replied:
Thanks for the explanation. Let's change it then.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -211,11 +255,22 @@ EXTRA_ARGS="
 
+
+if [ "$FSDP" -eq 1 ]; then
+EXTRA_ARGS="$EXTRA_ARGS --use-torch-fsdp2"
```
nit: indent this line.

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -211,11 +255,22 @@ EXTRA_ARGS="
+        echo "Warning: SEQ_PARALLEL cannot be used with FSDP."
+    fi
+else
+EXTRA_ARGS="$EXTRA_ARGS --use-distributed-optimizer --overlap-param-gather"
```
nit: indent this line.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -209,11 +243,21 @@ EXTRA_ARGS="
 "
 
+if [ "$FSDP" -eq 1 ]; then
+EXTRA_ARGS="$EXTRA_ARGS --use-torch-fsdp2"
```
nit: indent this line.

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -209,11 +243,21 @@ EXTRA_ARGS="
+        echo "Warning: SEQ_PARALLEL cannot be used with FSDP."
+    fi
+else
+EXTRA_ARGS="$EXTRA_ARGS --use-distributed-optimizer --overlap-param-gather"
```
nit: indent this line.

ryang-amd committed:
Add arguments for rope_fusion and data type, modify logic for fsdp with tp, sq and adam opti

ryang-amd commented:
Hi @wenchenvincent,
- I added two new arguments for ROPE_FUSION and DATA_TYPE 
- Update the logic among FSDP, TP, SQ and selection of optimizer when FSDP is off.
- Update the examples of corresponding arguments in README

Please let me know if there are some other modifications needed.

gurpreet-dhami reviewed (commented):
@ryang-amd 
No need to remove this - DATA_PATH=${DATA_PATH:-"$DATA_DIR/bookcorpus_text_sentence"}

We can always override the data path through arguments. Right ?

gurpreet-dhami reviewed (changes requested):
@ryang-amd
No need to remove this - DATA_PATH=${DATA_PATH:-"$DATA_DIR/bookcorpus_text_sentence"}

We can always override the data path through arguments. Right ?

ryang-amd committed:
change data path back to original one. Clarify on warning descriptions

ryang-amd committed:
change data path back for llama3.sh and some formating change

ryang-amd commented:
> @ryang-amd No need to remove this - DATA_PATH=${DATA_PATH:-"$DATA_DIR/bookcorpus_text_sentence"}
> 
> We can always override the data path through arguments. Right ?

Yes, changed them back accordingly.

ryang-amd committed:
update dataset preparation code to enable downloading wiki data, update instructions in readme

ryang-amd commented:
@wenchenvincent @gurpreet-dhami 
May I ask why there are duplicated lines (line 14-15, line17-18) in [prepare_dataset.sh#L17](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L17)? I assume it is typo. I removed it in the latest script.

ryang-amd commented:
Hey @wenchenvincent @gurpreet-dhami ,
I also update the dataset preparation code, so that user can choose to download either wiki data or bookcorpus data. 
The instructions in README is updated accordingly.

Please have a look when you have time.

gurpreet-dhami commented:
@ryang-amd , @wenchenvincent : shall we merge train_llama2 and train_llama3 files  ? 
Is there any other difference than model parameters ?

wenchenvincent commented:
> @ryang-amd , @wenchenvincent : shall we merge train_llama2 and train_llama3 files ? Is there any other difference than model parameters ?

Let's not address this issue in this PR. We can consolidate that if possible in another PR.

ryang-amd committed:
modify bookcorpus dataset downloading ode (a suspected bug) and update reaadme

ryang-amd commented:
Hey @wenchenvincent @gurpreet-dhami,
I made two changes:
1. I've updated one line in downloading the bookcorpus dataset because it will not download properly. The modification of downloading bookcorpus dataset could be revert to original version, if the downloading issue is fixed. (See below)
2. Updated readme file. 

My colleague @nsakkine found a potential bug when using the tools/preprocess_data.py to download bookcorpus dataset. Basically when using the default value of `--partitions` (partitions=1), it will not process the data correctly (empty .bin file), it looks like the following line causes the problem. Do you know why the line was written like this/who should we contact to fix that?
https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/tools/preprocess_data.py#L347-L348

wenchenvincent commented:
> @wenchenvincent @gurpreet-dhami May I ask why there are duplicated lines (line 14-15, line17-18) in [prepare_dataset.sh#L17](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L17)? I assume it is typo. I removed it in the latest script.

@gurpreet-dhami authored these I think. @gurpreet-dhami could you check?

wenchenvincent commented on examples/llama/train_llama2.sh:
```diff
@@ -76,6 +76,12 @@ CONTI_PARAMS="${CONTI_PARAMS:-0}"
 MCORE="${MCORE:-1}"
+OPTIMIZER="${OPTIMIZER:-adam}"
+FSDP="${FSDP:-1}"
+RECOMPUTE="${RECOMPUTE:-1}"
```
Can we change the default of recompute to 0?

wenchenvincent commented on examples/llama/train_llama3.sh:
```diff
@@ -74,17 +74,31 @@ CONTI_PARAMS="${CONTI_PARAMS:-0}"
+FSDP="${FSDP:-1}"
+RECOMPUTE="${RECOMPUTE:-1}"
+ROPE_FUSION="${ROPE_FUSION:-1}" # 1: use rope-fusion, 0: no-rope-fusion
+DATA_TYPE="${DATA_TYPE:-real}" # mock: use mock data, real: use real data
```
The name `DATA_TYPE` could be confusing. Can we use `MOCK_DATA` to control it?

wenchenvincent commented:
> Hey @wenchenvincent @gurpreet-dhami, I made two changes:
> 
> 1. I've updated one line in downloading the bookcorpus dataset because it will not download properly. The modification of downloading bookcorpus dataset could be revert to original version, if the downloading issue is fixed. (See below)
> 2. Updated readme file.
> 
> My colleague @nsakkine found a potential bug when using the tools/preprocess_data.py to download bookcorpus dataset. Basically when using the default value of `--partitions` (partitions=1), it will not process the data correctly (empty .bin file), it looks like the following line causes the problem. Do you know why the line was written like this/who should we contact to fix that?
> 
> https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/tools/preprocess_data.py#L347-L348

That line of code is from NV upstream: https://github.com/ROCm/Megatron-LM/blame/dea104b977b34f30b3b4a6d003352601b8721c92/tools/preprocess_data.py#L347-L348

ryang-amd committed:
modify 2 arguments and update the readme

ryang-amd commented:
Hi @wenchenvincent @gurpreet-dhami,

Small changes have been made according to Wen's latest comments:
- recompute=0 by default
- change DATA_TYPE to MOCK_DATA
- update README accordingly

For the bookcorpus dataset downloading issue with `--partitions=1`, we can keep the updated code where I added `--partitions=2` as an additional argument to preprocess the dataset, so that user can download the dataset without errors. We don't know yet why `partitions=1` doesn't work and it works only when running [this line ](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L14) twice. That's also the reason why there are duplicated lines in [prepare_dataset.sh](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L17). (Thanks @gurpreet-dhami for today's discussion!)

wenchenvincent commented:
> Hi @wenchenvincent @gurpreet-dhami,
> 
> Small changes have been made according to Wen's latest comments:
> 
> * recompute=0 by default
> * change DATA_TYPE to MOCK_DATA
> * update README accordingly
> 
> For the bookcorpus dataset downloading issue with `--partitions=1`, we can keep the updated code where I added `--partitions=2` as an additional argument to preprocess the dataset, so that user can download the dataset without errors. We don't know yet why `partitions=1` doesn't work and it works only when running [this line ](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L14) twice. That's also the reason why there are duplicated lines in [prepare_dataset.sh](https://github.com/ROCm/Megatron-LM/blob/dea104b977b34f30b3b4a6d003352601b8721c92/examples/llama/prepare_dataset.sh#L17). (Thanks @gurpreet-dhami for today's discussion!)

Thanks for addressing those. Could you also address the question whether FSDP and TP could be used together?

ryang-amd committed:
make FSDP and TP mutually exclusive(not using together)

ryang-amd committed:
update readme with pytorch version>=2.5

wenchenvincent reviewed (approved):
Thanks for the iterations of updates. LGTM.

gurpreet-dhami committed:
remove obsolete code for rope enablement
